### PR TITLE
Validation and parser transformations

### PR DESCRIPTION
**Pull request by ev1stensberg**
This PR is the initial work on the transformations. @pksjce you can add your work at the respective folders or tweak it as you want to. I suggest that @okonet and @pksjce is starting on the transformations, while me and @eliperelman integrate webpack-chain, yeoman and get our addon system viable. Let's make a final push for first release, it's not far away if we put some work into it the next couple of weeks.

### CODE REVIEW

File: lib/parser/init-transform.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+const fs = require('fs');
+const path = require('path');
+const transform = require('./transformations').transform;
+
+module.exports = function initTransform(opts) {
+	global.options = opts;
+	const templateFP = path.join(__dirname, 'utils', 'template.js');
+	const template = fs.readFileSync(templateFP, 'utf8');
+	const ast = transform(template);
```
Comment (okonet) [inline lib/parser/init-transform.js]: This is confusing since you'll get the new source and not the AST

File: lib/parser/init-transform.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	try {
+		const logicPath = process.cwd() + '/webpack.config.js';
+		fs.writeFileSync(logicPath, ast, 'utf8');
```
Comment (okonet) [inline lib/parser/init-transform.js]: ... and this made me think for a couple of seconds by looking at this. Let's rename `ast` to something like `new Source`

File: lib/parser/init-transform.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+const fs = require('fs');
+const path = require('path');
+const transform = require('./transformations').transform;
+
+module.exports = function initTransform(opts) {
+	global.options = opts;
+	const templateFP = path.join(__dirname, 'utils', 'template.js');
+	const template = fs.readFileSync(templateFP, 'utf8');
```
Comment (okonet) [inline lib/parser/init-transform.js]: It looks like you never remove the tmp file. You might want to take a look at https://www.npmjs.com/package/tmp to automate this.

File: lib/transformations/__snapshots__/utils.test.js.snap [hunk 4]
```diff
@@ -1,24 +1,8 @@
...
-exports[`utils createLiteral should create basic literal 1`] = `
-Object {
-  "comments": null,
-  "loc": null,
-  "regex": null,
-  "type": "Literal",
-  "value": "strintLiteral",
-}
-`;
+exports[`utils createLiteral should create basic literal 1`] = `"\\"stringLiteral\\""`;
```
Comment (okonet) [inline lib/transformations/__snapshots__/utils.test.js.snap]: Hmmm, seems like this is from another commit and doesn't relate to this PR. Do you mind rebasing it?

File: lib/parser/init-transform.js (hunk 3 above)
Comment (ev1stensberg) [inline lib/parser/init-transform.js]: I'm aware of this :) Thanks for the heads-up

File: lib/parser/init-transform.js (hunk 1 above)
Comment (ev1stensberg) [inline lib/parser/init-transform.js]: Is the initial work, I want @pksjce to work on this, I'm not fluent enough in writing AST's yet.

File: lib/parser/init-transform.js (hunk 2 above)
Comment (ev1stensberg) [inline lib/parser/init-transform.js]: Sure

File: lib/transformations/__snapshots__/utils.test.js.snap (hunk 4 above)
Comment (ev1stensberg) [inline lib/transformations/__snapshots__/utils.test.js.snap]: Sure!

File: lib/inquirer-prompt.js [hunk 5]
```diff
@@ -18,11 +18,14 @@ module.exports = function prompt(questions, config) {
 	.reduce(function (newOpts, ans) {
 		return attachAnswers(newOpts, ans);
 	}, Object.assign(config, require('./utils/initial-config')))
-	.flatMap(newOpts => {
+	.map(function(newOpts) {
```
Comment (ev1stensberg) [inline lib/inquirer-prompt.js]: since we used function at reduce, I figured out we should stick with the convention

Comment (okonet) [inline lib/inquirer-prompt.js]: Hmm, I thought the convention is to use `function` only for declarations, not for closures :/ Can you point me to the other usage?

File: lib/parser/init-transform.js (hunk 3 above)
Comment (okonet) [inline lib/parser/init-transform.js]: 👍

File: lib/inquirer-prompt.js (hunk 5 above)
Comment (ev1stensberg) [inline lib/inquirer-prompt.js]: I'll change it if so, didn't know that was the rule :)

**Comment by ev1stensberg** [comment]
Okay, done with what I think needs to be addressed first @okonet.  I think that we further need to discuss the transformations and how to use that effectively, with some of the current rules if doable. Let's bring it up tomorrow at the meeting :)

**Comment by ev1stensberg** [comment]
@okonet readfile didn't make validations work. I guess we can try change this a bit or try to validate after writing?

**Comment by ev1stensberg** [comment]
Closing this PR for now, maybe reopening later. We should first:

1. Add yeoman support
2. Hook the generators up with our internals for validation and attachments of objects to webpack-chain
3. Validate the config

**Pull request closed without merging**